app/controllers/meetings/edit/intro.js

```diff
@@ -0,0 +1,63 @@
+import Controller from '@ember/controller';
+import { task } from 'ember-concurrency';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+import { inject as service } from '@ember/service';
+
+export default class MeetingsEditIntroController extends Controller {
+  @service router;
+  @tracked editor;
+  @service intl;
+
+  get dirty() {
+    return this.model.intro !== this.editor.htmlContent;
+  }
+
+  @action
+  initEditor(editor) {
+    this.editor = editor;
+  }
+
+  @action
+  closeModal() {
+    this.router.transitionTo('meetings.edit', this.model.id);
+  }
+
+  @action
+  cancel() {
+    this.closeModal();
+  }
+  @action
+  async saveAndQuit() {
+    await this.saveTextTask.perform();
+    this.closeModal();
+  }
+
+  @task
+  *saveTextTask() {
+    const zitting = this.model;
+    zitting.intro = this.editor.htmlContent;
+    yield zitting.save();
+  }
+
+  addExitHandler() {
+    this.router.on('routeWillChange', this, this.confirm);
```

Comment:
I've done something similar in the past and instead of hooking this up in all the controllers I created a component to do it for me. Back then I kept it generic so I went for `<WillTransition>` but I think a more opinionated component might be better since this is one of  the only use-cases I can think of (and that would remove some extra setup code as well).

An API similar to this might be nice?

```hbs
<ConfirmRouteLeave @enabled={{this.dirty}} @message={{t "meetings.edit.intro.confirmQuitWithoutSaving"}}  />
```

That way the controllers would only need to store the dirty state and all the other setup code can be removed. I _think_ that should work unless I'm missing something.

Reply:
i almost did this as well, but I found it a weird usecase for a component. So I tried with a helper, but that didn't have a destructor, so I couldn't unregister the listener. And then I gave up because I was going mad

Reply:
Why do you think it's a weird use-case for it? Hmm, a class base helper should have a `willDestroy` hook, but I can't find it in the API docs. Not sure if that's a bug in the docs or not.

Reply:
Well it doesn't render anything
so the template is essentially a useless file
so id want a component without a template
which to me is basically a class based helper
but yeah I didnt see the hook in the docs, so I gave up there

Reply:
also incidentally, if it does have a willDestroy hook, how do you know this? I mean I've accepted you are an ember wizard, but where did you get your powers? :stuck_out_tongue:

Reply:
I've used it myself :smile: (which doesn't mean it's public API, I know), but I've seen official snippets where they use it like that. Which makes me assume it's simply a documentation bug. I'll try to find an official reference later.

Reply:
Well the same can be said for helpers that don't return a value :smile:. I think the framework just doesn't cover that use-case yet. "effects" _might_ solve it in the future but that's not implemented yet. Personally I don't have issues with components that don't really use the template. They can also simply serve as reusable logic. But maybe I'm biased because I've done it like that already.

Anyway I still think it's better than repeating this logic in all the controllers that need it :smile:. (I think the "new meeting" page might also want to use this in the future for example).

Reply:
yeah I think ill go for the component approach
rather a bit of a weird component than an undocumented helper hook I think

Reply:
alright done
a lot better for sure

Reply:
https://github.com/emberjs/rfcs/blob/master/text/0626-invoke-helper.md I found this RFC which includes a class based helper with a `willDestroy` hook, btw. There are probably more references, but that's the first "official" one I encountered :smile:. I also created an issue to update the API docs.